### PR TITLE
System display: T2564 Extend VyOS to support appliance LCDs

### PR DESCRIPTION
**Pull request by fmertz**
This is the implementation for Task [T2564](https://phabricator.vyos.net/T2564)

Thanks for considering it for inclusion in a future release.

### CODE REVIEW

**Comment by c-po** [comment]
Hi @fmertz,

thank you for the contribution. This happens to be the first non-maintainer upload using the new `get_config_dict()` implementation. Congratulations! 🎉 

You have almost followed our development guidelines, did you also happen to check out the contribution documentation? https://docs.vyos.io/en/latest/contributing/development.html#submit-a-patch

Unfortunately all your git commits consist out of the same headline which is not "that good". You might wan't to interactive rebase your changed `git rebase -i HEAD~4` and reword the commit messages to better describe your changes. I will commend inline with the GitHub review process.

Please note: I'm definately for adding this implementation but CLI design tends to be tricky. I also did not yet mount my LCD display into my APU4 development system to check your implementation but I will do this next week.

Is it possible for lcdproc to launch the deamon when there is no display attached? Like a dummy? 

In the meantime do you mind also providing a smoketest? We did not focus on this in the past but as we migrate more and more code oder to XML and python - and also rewriting this again in favour of get_config_dict() an embedded test is always preferred. You can make yourself comfortable with this small snipped used for SSH: https://github.com/vyos/vyos-smoketest/blob/master/scripts/cli/test_service_ssh.py

Thanks,
Christian

File: interface-definitions/system-display.xml.in
Comment (c-po) [inline interface-definitions/system-display.xml.in]: can we probably change this to lowercase please? 99% of the CLI use lowercase wording.

File: interface-definitions/system-display.xml.in
Comment (c-po) [inline interface-definitions/system-display.xml.in]: I do not get the intention of this node here. So I can configure the display and then turn it off? If this is the case, we usually use a `disable` leafNode which is `<valueless/>`. We tend to no longer add in new boolean nodes which are enable/disable.

File: interface-definitions/system-display.xml.in
Comment (c-po) [inline interface-definitions/system-display.xml.in:62]: I'm not yet confident if we really should expose "everything" that is configurable to the user - this will just add additional complexity. Why don't we decide which should be displayed?

**Comment by fmertz** [comment]
> This happens to be the first non-maintainer upload using the new `get_config_dict()` implementation. Congratulations! 🎉

Thanks for the kind words, it seemed like a natural way to go. The tricky parts were to guard against undefined variables and multi nodes with just one option.

>  `git rebase -i HEAD~4` 

I did this locally and squashed all 4 commits together (the last 3 commits were quick fixes and not worth evaluating separately). I then pushed force to my branch. I hope this makes it easier.

> LCD display into my APU4

Let me know the details of your LCD so I can add it to the  "model" options to help with testing. We can always take it out before the final merge if this is not a popular option.

> Is it possible for lcdproc to launch the deamon when there is no display attached? Like a dummy?

Picking the SDEC model invokes the sdeclcd driver, which issues direct parallel port commands. These do not fail if no LCD is connected, but you can check that the conf files get rendered consistently with the CLI (/etc/LCDd.conf and /etc/lcdproc.conf).

The other answer is that there is the "curses" driver that shows a pretend LCD as a window on the screen. If you start the LCDd server in the foreground, manually, in a separate session, it starts fine. Then the client options can be tested with the results showing on screen. This involves modifying the Python code so LCDd is not stopped/restarted, only lcdproc is.

> In the meantime do you mind also providing a smoketest?

I will have a look. Is there a write-up on what to do with this package once it is built? I am not setup for full ISO builds yet.

Thanks

File: interface-definitions/system-display.xml.in
Comment (c-po) [reply interface-definitions/system-display.xml.in:62]: Also I'm missing the VyOS version information here - this is sometimes very useful if there is a remote guy and tells you "device is malfunctioning" but he can't tell us the version.

File: src/conf_mode/system-display.py
Comment (c-po) [inline src/conf_mode/system-display.py]: This could be written much easier.

```
c = Config()
# we recommend using a list over a string
base = ['system', 'display']
# if service is not defined an empty dict is returned
lcd = c.get_config_dict(base)
# probably merge in default values here from XML tree (`<defaultValue>`)
return lcd

File: src/conf_mode/system-display.py
Comment (c-po) [inline src/conf_mode/system-display.py]: We should follow a common pattern which is `if not 'model' in c.keys()`

File: src/conf_mode/system-display.py
Comment (c-po) [inline src/conf_mode/system-display.py]: If you stop the entire service - or reconfigure it - `lcdproc` is always stopped, why not move this to the top of the function so it's always executed?

File: src/systemd/lcdproc.service
Comment (c-po) [inline src/systemd/lcdproc.service]: Please render file to the volatile location: `/run/lcdproc` so it won't be persistent.

**Review by c-po — CHANGES_REQUESTED** [review]

**Comment by c-po** [comment]
> I will have a look. Is there a write-up on what to do with this package once it is built? I am not setup for full ISO builds yet.

Simply PR the changes will automatically add it to the next ISO build. For vyos-smoketest you can "develop on target" and then simply copy over the python script to vyos-smoketest repo.

File: interface-definitions/system-display.xml.in
Comment (fmertz) [reply interface-definitions/system-display.xml.in]: OK, will do.

File: interface-definitions/system-display.xml.in
Comment (fmertz) [reply interface-definitions/system-display.xml.in]: What you are proposing was my intent. I was not aware of the valueless tag. I will do this.

File: src/conf_mode/system-display.py
Comment (fmertz) [reply src/conf_mode/system-display.py]: OK, will do.

File: src/conf_mode/system-display.py
Comment (fmertz) [reply src/conf_mode/system-display.py]: OK, will do.

File: src/conf_mode/system-display.py
Comment (fmertz) [reply src/conf_mode/system-display.py]: OK, will do.

File: src/systemd/lcdproc.service
Comment (fmertz) [reply src/systemd/lcdproc.service]: OK, I could do this, but the lcdproc Debian package already has /etc/LCDd.conf as well as /etc/lcdproc.conf. FWIW, the Debian package also has LCDd.service, and it runs LCDd with /etc/LCDd.conf.

So, I suppose we can follow the Debian "flavor" and leave everything in /etc or we can follow the VyOS "flavor" and render both files in /run.

Thanks for letting me know which way to go.

File: interface-definitions/system-display.xml.in
Comment (fmertz) [reply interface-definitions/system-display.xml.in:62]: I suppose we can decide on these options a bit later when you get to try them on your own LCD.

For the version, are you asking to show the VyOS version on the LCD somehow?

**Comment by c-po** [comment]
I have just noticed that the LCD pin header is a 2mm pitch and my mainboard has a 2.54mm pitch! awesome - need to somehow hack around it first.

**Comment by fmertz** [comment]
FWIW, I have been able to work around something like this before by forcing an insulated wire over the pin directly. This was for obtaining a serial connection to an onboard command prompt on an hard drive with a buggy firmware. The pin wedged itself inside the insulation and provided enough contact with the wire to work. Worth a try in a pinch.

Best of luck.

**Comment by jjakob** [comment]
I have a few Watchguard XTM5's laying around here, I can test it on one of them (IIRC they use sdeclcd driver with up/down/left/right buttons)

File: src/systemd/lcdproc.service
Comment (jjakob) [reply src/systemd/lcdproc.service]: The practice is to put all rendered configs in /run/$servicename, as they don't need to persist across reboot (/etc is intended for persistent configuration by the user/administrator with a few exceptions like hosts, resolv.conf). Then specify RuntimeDirectory and WorkingDirectory in the service file to point to that /run/$servicename directory.

**Comment by jjakob** [comment]
@c-po de-pinning the connector and putting a bit of insulating sleeving (shrink tube or kapton tape) over each pin, then plugging each pin in by hand is possible if the pin spacing allows for it. Otherwise I'd find a matching connector and make a adaptor.

Edit: de-pinning is possible only on individually pinned connectors, not on IDC connectors

**Comment by c-po** [comment]
Okay after a wiring issue I have a display connected and will test today. It tuns out my display seems to have some sort of an EEPROM as it can store test persistently

**Comment by fmertz** [comment]
Let me know if you want me to add your display "model" to this PR, even for just the testing phase. Let me know the specifics you need in LCDd.conf (driver name and any specifics in the driver's own section).

**Comment by c-po** [comment]
I'm testing using a 16x2 Character Display Module Crystalfontz Part Number: CFA533

**Comment by c-po** [comment]
```
cpo@LR4.wue3# set system display model sdec
[edit]
cpo@LR4.wue3# commit
[ system display ]
Display model is [REQUIRED]

[[system display]] failed
Commit failed
```
I also miss the part where I can specify the USB serial port used for connecting to the LCD assembly - in my case it would be usb0b1.1p1.0

```
cpo@LR4.wue3:~$ show system usb serial
Device           Model               Vendor
------           ------              ------
usb0b1.1p1.0     Crystalfontz_CFA533-USB_LCD Future Technology Devices International, Ltd
usb0b1.3p1.0     MC7710              Sierra Wireless, Inc.
usb0b1.3p1.2     MC7710              Sierra Wireless, Inc.
usb0b1.3p1.3     MC7710              Sierra Wireless, Inc.
```

**Comment by fmertz** [comment]
Can you try a local ~/LCDd.conf as follows and then try to start the server manually with "sudo LCDd -f -c ~/LCDd.conf"? The built-in server should show. If you need to change any of these settings, let me know and I will add them.

```
[server]
Driver=CFontzPacket
DriverPath=/usr/lib/x86_64-linux-gnu/lcdproc/

[CFontzPacket]
Model=533
Device=/dev/serial/by-bus/usb0b1.1p1.0
Contrast=350
Brightness=1000
OffBrightness=50
Reboot=yes
USB=yes
```

As we talked about early-on, we were trying not to expose every little setting from LCDd, and focus instead on the "model" being a macro group of settings, suitable for existing network appliances. Your situation for testing is of course different. I can create a test model with this group of settings once confirmed working.

I am not sure why there is that [REQUIRED] error on the model when you clearly specified it. I am running VyOS 1.3-rolling-202006221608 and my system-display branch is off of commit `acb36882e1ea5d5a5ec8a067ebde0f84f295c98e` and it works fine. Do I need to rebase the branch off of the latest?

Here:
```
vyos@vyosdev:~$ configure
[edit]
vyos@vyosdev# set system display model sdec 
[edit]
vyos@vyosdev# commit
[ system display ]
Display show must have a clock, host or network

[[system display]] failed
Commit failed
[edit]
vyos@vyosdev# 
```

Thanks for any feedback.

**Comment by fmertz** [comment]
Took the latest versions of ISO and vyos-1x, applied my code as a patch and the `Display model is [REQUIRED]` error can now be reproduced.

It seems `Config().get_config_dict(['system', 'display'])` now returns a dictionary at the 'display' level, not the lower level as before...

**Comment by c-po** [comment]
Please use get_first_key(). https://github.com/vyos/vyos-1x/blob/current/src/conf_mode/interfaces-pppoe.py#L53

**Comment by c-po** [comment]
I will now test using this and your suggestion:

Please also use a more "common" style when retrieving the dict.
```
def get_config():
    # Return a (possibly empty) configuration dictionary
    conf = Config()
    base = ['system', 'display']
    display = conf.get_config_dict(base, key_mangling=('-', '_'), get_first_key=True)
    # are there default values from the CLI that could be passed into dict?
    # see defaults()
    return display
```

```
set system display show host uptime
set system display model sdec
```


And then loading lcdproc gives:

```
3# sudo LCDd -f -c ~/LCDd.conf
LCDd 0.5.9, LCDproc Protocol 0.3
Part of the LCDproc suite
Copyright (C) 1998-2017 William Ferrell, Selene Scriven
                        and many other contributors

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software Foundation,
Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.

Could not open driver module CFontzPacket.so: CFontzPacket.so: cannot open shared object file: No such file or directory
Driver [CFontzPacket] binding failed
Could not load driver CFontzPacket
There is no output driver
Critical error while initializing, abort.
```

**Comment by fmertz** [comment]
In the [server] section, please add:

`DriverPath=/usr/lib/x86_64-linux-gnu/lcdproc/`

**Comment by c-po** [comment]
I also needed to change the device file path to `/dev/serial/by-bus/usb0b1.1p1.0`

A stupid question - can't those displays be autodetected like the type/model? Mine reports "Crystalfontz_CFA533-USB_LCD" which is a fair indicator - what about lanner? Then we can have a lookup table of supported modules and configure them "automagically"?

**Comment by fmertz** [comment]
> can't those displays be autodetected like the type/model?

It is not impossible, but it would have to be new code. It would have to be a combination of true detection (like this USB scenario) and guesses as some device cannot really be "detected", like the SDEC device on a parallel port. Guesses would have to be made, like Ethernet MAC address range (OUI?), type of ICH, etc, which would likely require access to the hardware for development. The SDEC driver works on a number of models, some likely hard to find at this point.

I was hoping to keep this simple for the users by offering only a limited number of "models", each with a nice help text listing all the supported hardware.

**Comment by c-po** [comment]
Well, we know on which hardware we are running, thus this will be easy I guess. Also Crystalfontz USB devices can be autodetected. We probably should discuss this on Slack as it could be faster. Also what is your development hardware?

**Comment by fmertz** [comment]
My development hardware is primarily the Portwell EZIO-100. It is stand alone and I can hook it up to a VyOS Virtualbox VM running on an old Dell laptop via a USB/Serial adapter. Some details (with pictures) are here:

[ezio-driver-for-lcdproc](https://forum.netgate.com/topic/115071/ezio-driver-for-lcdproc)

The other device is a older Watchguard XTM510 with the built-in SDEC, but I need to bring it back to life. The driver was done some time back.

**Comment by fmertz** [comment]
Proposal:

For now, let's leave the ezio and sdec models as they are in this package. The help string is enough to guide users towards the right model as we are talking about appliances, each with a fixed LCD model from the factory.

Let's add a "usb" node to deal with USB devices. There would have to be a script to somehow "detect" the proper LCD model off of the device/vendor combination, and then configure the LCDd.conf file accordingly. We would have to start small as only this CF533 is testable for now, but could be expanded over time as users provide additional model/vendor combinations. As this is for Linux only, hopefully we can get away with reading /sys and we don't need a USB Python library.

Thoughts?

**Comment by c-po** [comment]
Okay - so what I would do if I'd implement that was to keep a very restircted amount of possible CLI options:

More options = more trouble. Also more confusion from the user side - thus I tend to always use the least amount of CLI options.

* `set system lcd device /dev/serial/byy-bus/...` - if it's an USB based device it probably can be autodetected using regular USB ids - the 533 I used showed up, it it's not usb it will be something different - we check for ezio sdec. I guss EZIO/SDLEC devices will use any other device in /dev so we can derive the driver from it.

* `set system lcd display [uptime|version|load|memory|network]` - implemented as multi node so user can select the amount of info shown on the display - no furhter options like bps/Bps - networking is bits per second and for additional info SNMP should be used) - it is still a small display.

I have never seen a network appliance  up to now having this fine grained control over a display. Some displays not even show the running SW version - which tends to ease debugging on a system crash.

To me it makes absolutel no sense to mimic almost every option from lcdproc back to the CLI. The CLI is a user abstraction designed to make its life easier (whenever it can - but yes there are still poorly designed CLI interfaces)

To me follwing the KISS (Keep It Simple and Stupid) principle is one of the main goals. It's better to now invest more time in the logic of e.g. autodetection then on redesigning the CLI later on.

**Comment by fmertz** [comment]
It seems going with the actual device would be a cause of confusion for the users. The various LCD support threads out there seem to mention port selection all the time, and is a source of confusion. SDEC would be /dev/lpt?, and EZIO would be /dev/ttyS1, not super explicit. People seem to know what hardware they run, though, and seem happy to pick from a list.

Current CLI:

```
    system display model (sdec|ezio|test)
    system display show host (cpu|cpu-all|cpu-hist|disk|load-hist|memory|proc|uptime)
                        network interface <intName> alias <alias>
                                units (bps|Bps|pps)
                        clock (big|mini|date-time)
                        title <name>
    system display time <s>
    system display hello <string>
    system display bye <string>
    system display disabled

```
If you want to remove most of the CLI, we could bring it down to this:

```
system lcd model (sdec|ezio|usb)
system lcd display (cpu|load|memory|uptime|clock)
            network interface <intName> alias <alias>
```

model usb would have the auto-detection code behind it, with only the CF533 for now, expandable over time

The time parameter would be hard-coded to 20 seconds

The title would show the VyOS version

**Comment by c-po** [comment]
Why you hang onto the model part? If it is true what you say then we can assume sdec on /dev/lpt (I doubt there are many parallel ports out there) and choose ezio for /dev/ttyS - usb is well, usb based

**Comment by c-po** [comment]
I have integrated your changes locally and now checking on the possibilities of LCDproc. Is there a possibility to turn of the squares  in the top line?

**Comment by fmertz** [comment]
Can you post a picture? There is a heartbeat feature that can be turned on or off. Usually in the corner.

**Comment by c-po** [comment]
Please see Slack Video

**Comment by c-po** [comment]
Please see Slack Video

**Pull request merged**
